feat: add module documentation for module developer (#114)
AmadeusITGroup/otter: docs/core/MODULE.md

## List of Otter official modules

The full list of official modules can be found in the [Documentation Home page](../README.md#available-packages-and-tools).

## How to create its own Otter module

### Module development

> As indicated in the [definition session](#definition), the basic needs of an Otter module is to expose an `ng-add` schematics to be able to add the module via the command `ng add <package-name>`.

Reviewer: ```suggestion
As indicated in the [definition section](#definition), an Otter module should expose a `ng-add` schematic to allow developers to install the module via the command `ng add <package-name>`.
```

> > **Note**: Information relative to the `ng-add` schematics is available [here]().

Reviewer 2: Missing link for `ng-add` docs

> As any Otter module is following the same requirement, the new module can request the installation of any other by using providing helpers that will trigger the `ng-add` schematics of the requested module.

Reviewer: ```suggestion
As all the Otter modules follow the same requirement, the new module can use provided helpers to trigger the installation of other Otter modules through the `ng-add` schematic of the requested module.
```

Basic example of dependency module installation:

```typescript
import { chain, noop, Rule, SchematicContext, Tree } from '@angular-devkit/schematics';
import { NgAddSchematicsSchema } from './schema';

/**
 * Add Otter MyModule to an Otter Project
 *
 * @param options
 */
export function ngAdd(options: NgAddSchematicsSchema): Rule {
  return async (_tree: Tree, context: SchematicContext) => {
    try {

>       // use dynamic import to properly exception if not an Otter project

Reviewer 2: Use dynamic import to properly raise an exception if it is not an Otter project.

      const depsInfo = getO3rPeerDeps(path.resolve(__dirname, 'path', 'to', 'package.json'));
      return chain([
        // optional custom action dedicated to this module
        doCustomAction(),
        options.skipLinter ? noop() : applyEsLintFix(),
        // install ngx-translate and message format dependencies
        options.skipInstall ? noop : install,
        // add the missing Otter modules in the current project
        ngAddPackages(depsInfo.o3rPeerDeps, { skipConfirmation: true, version: depsInfo.packageVersion, parentPackageInfo: `${depsInfo.packageName!} - setup` })
      ]);
    } catch (e) {
      // If the installation is initialized in a non-Otter application, mandatory packages will be missing. We need to notify the user
      context.logger.error(`[ERROR]: Adding MyModule has failed.
      If the error is related to missing @o3r dependencies you need to install '@o3r/core' to be able to use the MyModule package. Please run 'ng add @o3r/core' .
      Otherwise, use the error message as guidance.`);
      throw (e);
    }
  };
}
```

### Register your module as official module

> It could make sense that the developed module would be part of the modules proposed in the beginning of the Otter application.

Reviewer: ```suggestion
It could make sense that the developed module would be part of the modules proposed at the beginning of the Otter application.
```
(not sure about this one)

Author: Not sure neither but surely less than you so I took it :D

> To be able to provide the module at Otter installation, it will be necessary to create a Pull Request to the [Otter repository](https://github.com/AmadeusITGroup/otter) containing the following steps:

Reviewer: ```suggestion
To provide the module at Otter installation, it is necessary to create a Pull Request to the [Otter repository](https://github.com/AmadeusITGroup/otter) that follows the steps below:
```

1. Ask to the user if the module should be part of this installation:

```json5
// in packages/@o3r/core/schematics/ng-add/schema.json

{
  "$schema": "http://json-schema.org/schema",
  "$id": "ngAddSchematicsSchema",
  "title": "Add Otter library",
  "description": "ngAdd Otter library",
  "properties": {
    "enableMyModule": {
      "type": "boolean",
      "default": false,
      "description": "Installation of myModule",
      "x-prompt": "Activate my module doing some stuff in your new application ?"
    }
  }
}
```

2. Add the conditional installation to the `ng-add` process

```typescript
// in packages/@o3r/core/schematics/ng-add/index.ts

import { chain, noop, Rule } from '@angular-devkit/schematics';
import { NgAddSchematicsSchema } from './schema';

/**
 * Add Otter library to an Angular Project
 *
 * @param options
 */
export function ngAdd(options: NgAddSchematicsSchema): Rule {

  // add the module within the external packages list
  const externalPackagesToInstallWithNgAdd = Array.from(new Set([
    ...(options.enableMyModule ? ['@my/module'] : [])
  ]));

}
```

> 3. Register the module within [packageGroup]() to make the module upgrade triggered with the Otter Framework upgrade.

Reviewer 2: Missing link for `packageGroup`?

> 3. Register the module within [packageGroup](https://github.com/angular/angular-cli/blob/main/docs/specifications/update.md) to make the module upgrade triggered with the Otter Framework upgrade.

Reviewer: ```suggestion
3. Register the module within [packageGroup](https://github.com/angular/angular-cli/blob/main/docs/specifications/update.md) to trigger the module upgrade with the Otter Framework upgrade.
```
